File: LEVEL3/가장 먼 노드.java
import java.io.*;
import java.util.*;

class Solution {
    static boolean[] chk;
    static ArrayList<ArrayList<Integer>> arr = new ArrayList<>();
    static Queue<Integer> que = new LinkedList<>();
    
    public int solve() {
        int cnt = 0;
        
        que.add(1);
        chk[1] = true;
        
        while(true) {
            Queue<Integer> tmp = new LinkedList<>();
            
            while(!que.isEmpty()) {
                int cur = que.poll();
                
                for(int i : arr.get(cur)) {
                    if(!chk[i]) {
                        tmp.add(i);
                        chk[i] = true;
                    }
                }
            }
            
            if(tmp.isEmpty()) {
                break;
            }
            
            que.addAll(tmp);
            cnt = tmp.size();
        }
        
        return cnt;
    }
    
    public int solution(int n, int[][] edge) {
        chk = new boolean[n + 1];
        
        for(int i = 0; i <= n; i++) {
            arr.add(i, new ArrayList<>());
        }
        
        for(int i = 0; i < edge.length; i++) {
            arr.get(edge[i][0]).add(edge[i][1]);
            arr.get(edge[i][1]).add(edge[i][0]);
        }
        
        int answer = solve();
        
        return answer;
    }
}

File: LEVEL3/가장 먼 노드.cpp
#include<bits/stdc++.h>
using namespace std;

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;
    int max = 0;
    queue<int> que;
    bool visited[20001] = {false, };
    bool chk[20001][20001] = {false, };
    int d[20001] = {0, };
    
    sort(edge.begin(), edge.end());
    
    for(int i = 0; i < edge.size(); i++) {
        chk[edge[i][0]][edge[i][1]] = true;
        chk[edge[i][1]][edge[i][0]] = true;
    }
    
    que.push(1);
    visited[1] = true;
    
    while(!que.empty()) {
        int cur = que.front();
        que.pop();
        
        for(int i = 2; i <= n; i++) {
            if(chk[cur][i] && !visited[i]) {
                que.push(i);
                visited[i] = true;
                d[i] = d[cur] + 1;
                
                if(d[i] > max) {
                    max = d[i];
                }
            }
        }
    }
    
    for(int i = 1; i <= n; i++) {
        if(max == d[i]) {
            answer++;
        }
    }
    
    return answer;
}